### ISSUE TITLE
Slow down of a long-running program with OCaml 5.2

### ISSUE THREAD

**Issue by FardaleM** [ocaml/ocaml#13513]
To analyse the performance of the core procedure used in [Dowsing](https://github.com/Drup/dowsing/), I made a long-running program. This experiment is on the branch [hall_of_fame](https://github.com/Drup/dowsing/tree/hall_of_fame).

This does the following: given a list of 44618 types, for each type `t1`, `t1` is unified with every other type from the list. I will call the loop over `t1` the outer loop and the looping doing the unification against `t1`  the inner loop. After some time (around an hour), I noticed a massive slowdown, meaning that the inner loop does no longer finish in more than 30 min, while if I run only this loop, it takes less than a minute. Every unification should be independent, and therefore, there should be no impact from one step of the outer loop to the next one.

I am using OCaml 5.2 to run this experiment. Here is what I tried to understand and find the reason for this slow down.
- I put a `Gc.full_major` for each turn of the outer loop, I also tried with `Gc.compact` and this did not change anything.
- I printed the stats of the GC with `Gc.print_stat` after the compaction. The number of words is constant, so it does not seem like there is a memory leak.
- I could not try with the recent alpha of OCaml 5.3 as odoc is not yet compatible with it.

**Comment by gasche** [comment]
It could be that `Gc.print_stat` is wrong, and in fact you are leaking memory and swapping. Have you looked at the memory consumption reported by external tools such as `top` or `htop`?

Other question: do you observe the same behavior with OCaml 4.14?

**Comment by FardaleM** [comment]
I spent a bit of time exploring in more details this behaviour, but with the changes I made, the slowdown disappeared. There was a problem with the code that could explain the slowdown, but I am not 100% sure it was that. I will keep an eye on this to see if I can reproduce it again.

For testing with OCaml 4.14, unfortunately, the code is depending on `compiler_libs` and there is not really possible to test with 4.14.

I will close this issue and reopen it if I can reproduce the slowdown.
